### DISCUSSION TITLE
ステート管理「useContext」と「Redux」の使い分けに関して

### DISCUSSION THREAD

**Discussion by nabesuke** [FisaCloudAlliance/JamstackEducation#65] [Q&A]
ステート管理の「useContext」と「Redux」の使い分けについて質問です。

react hooksの登場する以前は、ステート管理出来なかったため、Reduxを利用していたかと思うのですが、
react hooks(useState、useContext)が実装され、ステート管理が可能となった現状でも、Reduxを利用したほうが良いのでしょうか？
セミナー後半＋Wikiに、Reduxの項もあったので、Reduxを利用したほうがベストなのか？と疑問に感じました。

よろしくお願い致します。

**Comment by s-ueno** [comment]
とても有意義なご質問、ありがとうございます！

react は 去年の暮れに、ver . 16 から　ver 17 にメジャーバージョンアップし、さらに 18 への更新もアナウンスされています。

この hooks も 2019年の大型アップデートでクラス指向のreact から 関数型指向の react へ 大きく進化しました。

Redux の仕組みは、コンパイラーが検知して自動で作成できるはずだから、今の書き方は改善したい旨がreactでアナウンスされています。

そうはいっても、まだ異なるコンポーネント間のやり取りや、クライアントにキャッシュして、次回起動時に復元する機能の恩恵を受けるためにも、Reduxはまだまだ必要です。

ただ、新しく出てきたソリューション、特に useState は関数型プログラミングを促進させる目的で登場したものなので、住み分けた上で　併用し、　新しい技術は積極的に取り入れる姿勢がデジタルディスラプション時代の技術者に求められていると思います。

**Redux**

- 主に参照系
- 親子関係などが成立しない、異なるコンポーネント間でのバインド連動
- 画面初期データ取得など、API経由でデータ取得するもの
  - ローカルに一時キャッシュすることで、ユーザーエクスペリエンスの工場が見込めるもの

**useState**

- 主に入力系
- 入力画面や検索画面など、ユーザーが都度入力した値の場合は、復元する必要がなく(ブラウザの入力補完もあり)、スコープも小さい方が安全性が高いため
